File: src/data_rover/core/model/naming.py
"""Display-name derivation shared by the API read routes and the IndexSet.

Kept in lock-step with the frontend's ``elementDisplayName``
(``frontend/src/lib/util/element-name.ts``) so a row's label is identical
whether it comes from the lite (server) or full (client) source. The roots
order index sorts by this value, so moving/changing it changes server-side
tree ordering — treat the semantics as frozen.
"""

from __future__ import annotations

from .element import Element


def _name_str(value: object) -> str | None:
    """A usable name out of a property value: the string itself, or — for a
    multiplicity-many ``name`` (a list, e.g. from a migrated legacy model) —
    the first non-empty string entry."""
    if isinstance(value, str) and value:
        return value
    if isinstance(value, (list, tuple)):
        for v in value:
            if isinstance(v, str) and v:
                return v
    return None


def display_name(element: Element) -> str:
    """The case-insensitive non-empty ``name`` property, else the id.

    An exact lowercase ``name`` wins over other casings (``Name``/``NAME``).
    List values (multiplicity-many names) contribute their first non-empty
    string entry.
    """
    props = element.properties
    exact = _name_str(props.get("name"))
    if exact is not None:
        return exact
    for key, value in props.items():
        if key != "name" and key.lower() == "name":
            found = _name_str(value)
            if found is not None:
                return found
    return element.id

File: engine/src/model/naming.ts
import type { Value } from '../value/types.ts';
import { getProp, type ElementRec } from './records.ts';

/** A non-empty string, or the first non-empty string of a list. */
function nameStr(value: Value | undefined): string | null {
	if (typeof value === 'string') return value === '' ? null : value;
	if (Array.isArray(value)) {
		for (const item of value) if (typeof item === 'string' && item !== '') return item;
	}
	return null;
}

/**
 * The element's `name`, or `null`. An exact lower-case `name` wins over any
 * other casing (`Name`, `NAME`), which are tried in property order.
 */
export function nameOf(element: ElementRec): string | null {
	const exact = nameStr(getProp(element.props, 'name'));
	if (exact !== null) return exact;
	for (const key of Object.keys(element.props)) {
		if (key !== 'name' && key.toLowerCase() === 'name') {
			const found = nameStr(element.props[key]);
			if (found !== null) return found;
		}
	}
	return null;
}

/** The element's name, else its id. Root order sorts by this. */
export function displayName(element: ElementRec): string {
	return nameOf(element) ?? element.id;
}
